### ISSUE TITLE
RFE: Extend unit section with custom journal fields

### ISSUE THREAD

**Issue by zeisss** [systemd/systemd#3041]
### Submission type
- [ ] Bug report
- [x] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

> systemd 229

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### Used distribution

> CoreOS
### RFE

The [journal-fields](https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html) man page states:

> New fields may freely be defined by applications, but a few fields have special meaning.

But, adding new fields only seem to be doable by accessing the journal api directly. We would like to specify this in the `.service` files so that all log output generated by the service (including the systemd messages like "Starting service ...") are annotated with new fields, e.g. customer ID, production/staging environment etc.

This could work like the `Environment` fields:

```
[Unit]
Description=My service
JournalField="CUSTOMER=1"
JournalField="ONEFIELD=2 ANOTHER=1234-12314123-15123-1245123"
```

**Background**: We (at Giant Swarm) are running a bunch of services on our servers for various customers. We also forward the log data from journald as json streams via logstash to ElasticSearch. We would like to be able to filter inside logstash the data for certain customers based on fields in the JSON data.

**Comment by jonboulle** [comment]
We'd also love to see this for use in rkt.

**Comment by lucab** [comment]
I had a look at how many moving parts are involved here and which need to be changed, and I think that an implementation plan may be as follow:
- add a `JournalField=` option to [properties parser](https://github.com/systemd/systemd/blob/v231/src/core/dbus-execute.c#L820), accepting single "K=V" UTF-8 strings
- pass newline-terminated "K=V" when [connecting the logger](https://github.com/systemd/systemd/blob/v231/src/core/execute.c#L292), plus a trailing underscore-prefixed header terminator
- add a new stdout_stream entry as the [penultimate state](https://github.com/systemd/systemd/blob/to-upstream/custom-journal-fields/src/journal/journald-stream.c#L285), in order to parse user fields
- store them linked to the [StdoutStream at hand](https://github.com/systemd/systemd/blob/v231/src/journal/journald-stream.c#L63)
- attach them to every final entry that will be spit out

I'm not really familiar with this code, so I'd like to hear some experienced opinions before touching things around. 

Moreover, I have an open concern related to namespacing: I guess nobody wants to see a `MESSAGE=foo` userfield floating around. Should keys be blacklist-matched upfront, or just namespaced? (I'm in favor of the latter, eg. transmuting to `SDUSER_MESSAGE=foo`)

**Comment by johannbg** [comment]
Adding or otherwise manipulating journal fields is not something that belongs in type services units. 
Any particular reason you aren't using _MACHINE_ID and _HOSTNAME which should be enought to be able to map the container(s) directly to customers in whatever logging backend you are using?

**Comment by zeisss** [comment]
@johannbg No, the hostname/machine-id is simply not unique enough if you are sharing multiple customers on one machine.

**Comment by johannbg** [comment]
If those customers are isolated into their own os containers it should be sufficient if not you can always add SyslogIdentifier= to type service to the mix as well like SyslogIdentifier=customer3

**Comment by zeisss** [comment]
@johannbg Ok, I think I got you know. Just to be clear: You suggest to start a container via `systemd-nspawn`, with it's own systemd/machine-id and run the docker clients in there. With this we can have a deterministic machine-id/hostname and identify the environment this way later.

Might work, but still needs more work / moving parts than just being able to extend a unit file with the attributes I need IMO.

Thanks for the suggestion :)

**Comment by johannbg** [comment]
Or just run systemd containers directly no need to add docker to the mix ;) 
Anyway if
 _MACHINE_ID ( which tells which machine )
_HOSTNAME ( which tells which host )
SYSLOG_IDENTIFIER ( which tells which customer )
UNIT ( which tells which unit is running )
is not enough for whatever reason to identify this then OS_RELEASE= and the content of os-release would need to be added to the journal but type service units should not be used to configure/contain these attributes nor having the ability to add custom journal entries I would think

then you could extend os-release to cover the business case and identification with something like
CUSTOMER_ID=1234-12314123-15123-1245123
CUSTOMER=ACKME
CONTAINER_ID=43
ROLE=WEBSERVER
etc...
and that would end up in the OS_RELEASE= entry in the journal at least that sound flexible enough as well as being better implementation than adding this to the type unit files.

**Comment by poettering** [comment]
@lucab so, i'd prefer a solution that would attach these fields also to log lines generated with syslog and the journal native logging. My current suggestion would hence be to attach this info to the cgroup of the service that is logging so that we can read it from there. One way to do this is by making use of xattrs in cgroupfs (which have been supported now for quite some time). Specifically, we'd define one xattr (maybe called "user.journal") or so, which contains a NUL-separated blob with all journal props to set for all log messages received from processes from that cgroup. PID 1 should write the fields set with JournalFields= onto the cgroup of the service at the time of creating it. And journald should read it from the cgroup at the time of recv of a msg (or in the case of the stdout/stderr stream: at the time of connection). In fact, it should probably not just read the journal fields from the cgroup itself, but also from all its parents, until the root is reached.

This all is a bit racy, unfortunately. This is because the cgroup info might not always be available anymore when we read a msg and get its PID. for the stream logging we can minimize this race, by caching the data in the connection. For the rest we probably have to accept it.

**Comment by lucab** [comment]
@poettering thanks for the pointers, I see the push for a more general solution (I was missing the other two usecases in my analysis). 

I just have a concern, as this seems to introduce quite a bit of cgroup-walking overhead to every logging operation, and I've seen multiple reporters concerned with logging performances. Maybe the discussion at https://github.com/systemd/systemd/pull/2280 should be restarted first?

/cc @vcaputo

**Comment by vcaputo** [comment]
The raciness is not even limited to just the cgroup info not being available anymore, even if that were guaranteed to accompany every log entry much like the ucreds do, you still go and sample the rest of the information keyed off the cgroup and pid at a time that is in the future relative to when the log message was actually produced.

The discussion in #2280 touches on this subject, and the current implementation of _always_ sampling the metadata from the pseudo filesystems at the time of every log line being processed is not even attempting to mitigate the effects of the raciness.  As I explain in that PR, we should only sample the associated information once for every chunk/buffer of log data consumed, not for every line.

I don't see how we can eliminate all the raciness from this without getting _all_ the metadata sampled atomically by the kernel at the time the log gets produced, and attached to the buffer containing the in-flight log data.  Perhaps that's where we should be focusing our time instead?  Maybe there needs to be some new form of IPC where a logging daemon like journald can set aside memory for storing such metadata and instructing the kernel to capture it into that memory atomically when a write occurs on an associated fd... thinking out loud here.

**Comment by poettering** [comment]
@lucab PR #4067 shows how attach xattrs to cgroups and adds all necessary infrastructure to it. it also reads that data from journald. With that in place it should be pretty straighforward to implement this feature as well.

I think we can leav walking the cgroup tree up out for now, and make this specific to leaf units (i.e. not slices)

**Comment by poettering** [comment]
And yes, we should discuss #2280 again. Need to think about it.

**Comment by poettering** [comment]
Fix waiting in #7246
